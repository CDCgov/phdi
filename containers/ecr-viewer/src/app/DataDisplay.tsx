import React, { ReactNode, useEffect, useState } from "react";
import { Button } from "@trussworks/react-uswds";
import classNames from "classnames";

import { ToolTipElement } from "@/app/ToolTipElement";

export interface DisplayDataProps {
  title?: string;
  className?: string;
  toolTip?: string;
  value?: string | React.JSX.Element | React.JSX.Element[] | React.ReactNode;
  dividerLine?: boolean;
}

/**
 * Functional component for displaying data.
 * @param props - Props for the component.
 * @param props.item - The display data item to be rendered.
 * @param [props.className] - Additional class name for styling purposes.
 * @returns - A React element representing the display of data.
 */
export const DataDisplay: React.FC<{
  item: DisplayDataProps;
  className?: string;
}> = ({
  item,
  className,
}: {
  item: DisplayDataProps;
  className?: string;
}): React.JSX.Element => {
  item.dividerLine =
    item.dividerLine == null || item.dividerLine == undefined
      ? true
      : item.dividerLine;
  return (
    <div>
      <div className="grid-row">
        <div className="data-title">
          <ToolTipElement content={item.title} toolTip={item.toolTip} />
        </div>
        <div
          className={classNames(
            "grid-col-auto maxw7 text-pre-line",
            className,
            item.className ? item.className : "",
          )}
        >
          <FieldValue>{item.value}</FieldValue>
        </div>
      </div>
      {item.dividerLine ? <div className={"section__line_gray"} /> : ""}
    </div>
  );
};
/**
 * Functional component for displaying a value. If the value has a length greater than 500 characters, it will be split after 300 characters with a view more button to view the entire value.
 * @param value - props for the component
 * @param value.children - the value to be displayed in the value
 * @returns - A React element representing the display of the value
 */
const FieldValue: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  const maxLength = 500;
  const cutLength = 300;
  const [hidden, setHidden] = useState(true);
  const [fieldValue, setFieldValue] = useState(children);
  const valueLength = getReactNodeLength(children);
  const cutField = trimField(children, cutLength, setHidden).value;
  useEffect(() => {
    if (valueLength > maxLength) {
      if (hidden) {
        setFieldValue(cutField);
      } else {
        setFieldValue(
          <>
            {children}&nbsp;
            <Button
              type={"button"}
              unstyled={true}
              onClick={() => setHidden(true)}
            >
              View less
            </Button>
          </>,
        );
      }
    }
  }, [hidden]);

  return fieldValue;
};
/**
 * Recursively determine the character length of a ReactNode
 * @param value - react node to be measured
 * @returns - the number of characters in the ReactNode
 */
const getReactNodeLength = (value: React.ReactNode): number => {
  if (typeof value === "string") {
    return value.length;
  } else if (Array.isArray(value)) {
    let count = 0;
    value.forEach((val) => (count += getReactNodeLength(val)));
    return count;
  } else if (React.isValidElement(value) && value.props.children) {
    return getReactNodeLength(value.props.children);
  }
  return 0;
};
/**
 * Create an element with `remainingLength` length followed by a view more button
 * @param value - the value that will be cut
 * @param remainingLength - the length of how long the returned element will be
 * @param setHidden - a function used to signify that the view more button has been clicked.
 * @returns - an object with the shortened value and the length left over.
 */
const trimField = (
  value: React.ReactNode,
  remainingLength: number,
  setHidden: (val: boolean) => void,
): { value: React.ReactNode; remainingLength: number } => {
  if (remainingLength < 1) {
    return { value: null, remainingLength };
  }
  if (typeof value === "string") {
    const cutString = value.substring(0, remainingLength);
    if (remainingLength - cutString.length === 0) {
      return {
        value: (
          <>
            {cutString}...&nbsp;
            <Button
              type={"button"}
              unstyled={true}
              onClick={() => setHidden(false)}
            >
              View more
            </Button>
          </>
        ),
        remainingLength: 0,
      };
    }
    return {
      value: cutString,
      remainingLength: remainingLength - cutString.length,
    };
  } else if (Array.isArray(value)) {
    let newValArr = [];
    for (let i = 0; i < value.length; i++) {
      let splitVal = trimField(value[i], remainingLength, setHidden);
      remainingLength = splitVal.remainingLength;
      newValArr.push(
        <React.Fragment key={`arr-${i}-${splitVal.value}`}>
          {splitVal.value}
        </React.Fragment>,
      );
    }
    return { value: newValArr, remainingLength: remainingLength };
  } else if (React.isValidElement(value) && value.props.children) {
    let childrenCopy: ReactNode;
    if (Array.isArray(value.props.children)) {
      childrenCopy = [...value.props.children];
    } else {
      childrenCopy = value.props.children;
    }
    let split = trimField(childrenCopy, remainingLength, setHidden);
    const newElement = React.cloneElement(
      value,
      { ...value.props },
      split.value,
    );
    return { value: newElement, remainingLength: split.remainingLength };
  }
  return { value, remainingLength: remainingLength };
};
