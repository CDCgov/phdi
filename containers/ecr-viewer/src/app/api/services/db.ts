import pgp from "pg-promise";

const db_url = process.env.DATABASE_URL || "";
export const database = pgp()(db_url);
