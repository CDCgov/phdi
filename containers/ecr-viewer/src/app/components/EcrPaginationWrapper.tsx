"use client";

import { Label, Select } from "@trussworks/react-uswds";
import React, { ReactNode, useEffect, useState } from "react";
import { Pagination } from "@/app/components/Pagination";
import { usePathname, useRouter, useSearchParams } from "next/navigation";

interface EcrPaginationWrapperProps {
  totalCount: number;
  children: ReactNode;
}

interface UserPreferences {
  itemsPerPage: number;
}

const defaultPreferences = {
  itemsPerPage: 25,
};

/**
 * Renders a list of eCR data with viewer.
 * @param props - The properties passed to the component.
 * @param props.totalCount - Total number of eCRs
 * @param props.children - Contents of the wrapper
 * @returns The JSX element (table) representing the rendered list of eCRs.
 */
const EcrPaginationWrapper = ({
  totalCount,
  children,
}: EcrPaginationWrapperProps) => {
  const searchParams = useSearchParams();
  const router = useRouter();
  const pathname = usePathname();

  const currentPage = Number(searchParams.get("page")) || 1;
  const [userPreferences, setUserPreferences] =
    useState<UserPreferences>(defaultPreferences);

  useEffect(() => {
    const userPreferencesString = localStorage.getItem("userPreferences");
    if (userPreferencesString) {
      setUserPreferences(JSON.parse(userPreferencesString));
    }
  }, []);

  useEffect(() => {
    const current = new URLSearchParams(Array.from(searchParams.entries()));
    current.set("itemsPerPage", userPreferences.itemsPerPage.toString());

    const search = current.toString();
    const query = search ? `?${search}` : "";
    router.push(`${pathname}${query}`);
  }, [userPreferences]);

  const totalPages = Math.ceil(totalCount / userPreferences.itemsPerPage);

  const startIndex = (currentPage - 1) * userPreferences.itemsPerPage + 1;
  const endIndex = Math.min(
    currentPage * userPreferences.itemsPerPage,
    totalCount,
  );

  return (
    <div className="main-container height-full flex-column flex-align-center">
      {children}
      <div className="pagination-bar width-full padding-x-3 padding-y-105 flex-align-self-stretch display-flex flex-align-center">
        <div className={"flex-1"}>
          Showing {startIndex}-{endIndex} of {totalCount} eCRs
        </div>
        <Pagination
          currentPage={currentPage}
          totalPages={totalPages}
          pathname={""}
          className={"flex-1"}
        />
        <div
          className={"display-flex flex-align-center flex-1 flex-justify-end"}
        >
          <Label
            htmlFor="input-select"
            className={"margin-top-0 margin-right-1025"}
          >
            eCRs per page
          </Label>
          <Select
            id="input-select"
            name="input-select"
            value={userPreferences.itemsPerPage}
            className={"styled width-11075 margin-top-0"}
            onChange={(e) => {
              const updatedUserPreferences: UserPreferences = {
                ...userPreferences,
                itemsPerPage: +e.target.value,
              };
              setUserPreferences(updatedUserPreferences);
              localStorage.setItem(
                "userPreferences",
                JSON.stringify(updatedUserPreferences),
              );
            }}
          >
            <React.Fragment>
              <option value="6">6</option>
              <option value="25">25</option>
              <option value="50">50</option>
              <option value="75">75</option>
              <option value="100">100</option>
            </React.Fragment>
          </Select>
        </div>
      </div>
    </div>
  );
};

export default EcrPaginationWrapper;
