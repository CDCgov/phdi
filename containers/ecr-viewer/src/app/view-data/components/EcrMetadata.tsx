import { DisplayData, ReportableConditions } from "../../utils";
import { Fragment } from "react";

import {
  AccordianSection,
  AccordianH4,
  AccordianDiv,
} from "../component-utils";
import { SectionConfig } from "./SideNav";
import { Table } from "@trussworks/react-uswds";

interface EcrMetadataProps {
  rrDetails: ReportableConditions;
  eicrDetails: DisplayData[];
  eCRSenderDetails: DisplayData[];
}

export const ecrMetadataConfig: SectionConfig = new SectionConfig(
  "eCR Metadata",
  ["RR Details", "eICR Details", "eCR Sender Details"],
);

interface ReportableConditionsList {
  [condition: string]: {
    [trigger: string]: Set<string>; // Maps a trigger to a set of locations
  };
}

const convertDictionaryToRows = (dictionary: ReportableConditionsList) => {
  if (!dictionary) return [];
  const rows: JSX.Element[] = [];
  Object.entries(dictionary).forEach(([condition, triggers], _) => {
    Object.entries(triggers).forEach(([trigger, locations], triggerIndex) => {
      const locationsArray = Array.from(locations);
      locationsArray.forEach((location, locationIndex) => {
        const isConditionRow = triggerIndex === 0 && locationIndex === 0;
        const isTriggerRow = locationIndex === 0;
        const conditionCell = isConditionRow ? (
          <td
            rowSpan={Object.keys(triggers).reduce(
              (acc, key) => acc + Array.from(triggers[key]).length,
              0,
            )}
          >
            {condition}
          </td>
        ) : null;
        const triggerCell = isTriggerRow ? (
          <td rowSpan={locationsArray.length}>{trigger}</td>
        ) : null;

        rows.push(
          <tr key={`${condition}-${trigger}-${location}`}>
            {conditionCell}
            {triggerCell}
            <td>{location}</td>
          </tr>,
        );
      });
    });
  });

  return rows;
};

/**
 * Functional component for displaying eCR metadata.
 * @param props - Props containing eCR metadata.
 * @param props.rrDetails - The reportable conditions details.
 * @param props.eicrDetails - The eICR details.
 * @param props.eCRSenderDetails - The eCR sender details.
 * @returns The JSX element representing the eCR metadata.
 */
const EcrMetadata = ({
  rrDetails,
  eicrDetails,
  eCRSenderDetails,
}: EcrMetadataProps) => {
  return (
    <AccordianSection>
      <AccordianH4>
        <span id={ecrMetadataConfig.subNavItems?.[0].id}>
          {ecrMetadataConfig.subNavItems?.[0].title}
        </span>
      </AccordianH4>
      <AccordianDiv>
        <Table bordered caption="Reportibility Summary" className="rrTable">
          <thead>
            <tr>
              <th className="reportability_summary_header">
                Reportable Condition
              </th>
              <th>RCKMS Rule Summary</th>
              <th className="reportability_summary_header">
                Jurisdiction Sent eCR
              </th>
            </tr>
          </thead>
          <tbody>{convertDictionaryToRows(rrDetails)}</tbody>
        </Table>
        <div className={"padding-bottom-1"} />
        <AccordianH4>
          <span id={ecrMetadataConfig.subNavItems?.[1].id}>
            {ecrMetadataConfig.subNavItems?.[1].title}
          </span>
        </AccordianH4>
        {eicrDetails.map(({ title, value }) => {
          return (
            <Fragment key={title}>
              <div className="grid-row">
                <div className="data-title">{title}</div>
                <div className="grid-col-auto text-pre-line">{value}</div>
              </div>
              <div className={"section__line_gray"} />
            </Fragment>
          );
        })}
        <div className={"padding-bottom-1"} />
        <AccordianH4>
          <span id={ecrMetadataConfig.subNavItems?.[2].id}>
            {ecrMetadataConfig.subNavItems?.[2].title}
          </span>
        </AccordianH4>
        {eCRSenderDetails.map(({ title, value }) => {
          return (
            <Fragment key={title}>
              <div className="grid-row">
                <div className="data-title">{title}</div>
                <div className="grid-col-auto text-pre-line">{value}</div>
              </div>
              <div className={"section__line_gray"} />
            </Fragment>
          );
        })}
      </AccordianDiv>
    </AccordianSection>
  );
};

export default EcrMetadata;
