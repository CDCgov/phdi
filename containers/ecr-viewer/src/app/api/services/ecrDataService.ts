import pgPromise from "pg-promise";
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { streamToJson } from "@/app/api/services/utils";
import { database } from "@/app/api/services/db";

const S3_SOURCE = "s3";
const POSTGRES_SOURCE = "postgres";
const s3Client = new S3Client({ region: process.env.AWS_REGION });

/**
 * Get eCR data based on the ecrID.
 * @param ecrId - The id of the eCR.
 * @returns A promise resolving to eCR data.
 */
export const getEcrData = async (ecrId: string) => {
  if (process.env.SOURCE === S3_SOURCE) {
    return await getS3Data(ecrId);
  } else if (process.env.SOURCE === POSTGRES_SOURCE) {
    return await getPostgresData(ecrId);
  }
};

/**
 * Retrieves FHIR data from Postgres based on eCR ID.
 * @param ecrId - The id of the ecr.
 * @returns A promise resolving to eCR data.
 */
const getPostgresData = async (ecrId: string) => {
  const { ParameterizedQuery: PQ } = pgPromise;
  const findFhir = new PQ({
    text: "SELECT * FROM fhir WHERE ecr_id = $1",
    values: [ecrId],
  });
  return (await database.one(findFhir)).data;
};

/**
 * Retrieves FHIR data from S3 based on eCR ID.
 * @param ecrId - The id of the ecr.
 * @returns A promise resolving to eCR data.
 */
const getS3Data = async (ecrId: string) => {
  const bucketName = process.env.ECR_BUCKET_NAME;
  const objectKey = `${ecrId}.json`; // This could also come from the request, e.g., req.query.key
  const command = new GetObjectCommand({
    Bucket: bucketName,
    Key: objectKey,
  });

  const { Body } = await s3Client.send(command);
  return await streamToJson(Body);
};
