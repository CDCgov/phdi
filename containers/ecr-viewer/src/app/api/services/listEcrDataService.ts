import {
  S3Client,
  ListObjectsV2Command,
  ListObjectsV2CommandOutput,
} from "@aws-sdk/client-s3";
import { database } from "@/app/api/services/db";
import {
  formatDateTime,
  convertUTCToLocalString,
} from "@/app/services/formatService";

const S3_SOURCE = "s3";
const POSTGRES_SOURCE = "postgres";
const s3Client = new S3Client({ region: process.env.AWS_REGION });

export type ListEcr = {
  ecrId: string;
  dateModified: string;
}[];

/**
 * Handles GET requests by fetching data from different sources based on the environment configuration.
 * It supports fetching from S3 and Postgres. If the `SOURCE` environment variable is not set to
 * a supported source, it returns a JSON response indicating an invalid source.
 * @returns A promise that resolves to a `NextResponse` object
 *   if the source is invalid, or the result of fetching from the specified source.
 *   The specific return type (e.g., the type returned by `list_s3` or `list_postgres`)
 *   may vary based on the source and is thus marked as `unknown`.
 */
export async function listEcrData() {
  if (process.env.SOURCE === S3_SOURCE) {
    const data = await list_s3();
    return processListS3(data);
  } else if (process.env.SOURCE === POSTGRES_SOURCE) {
    const data = await list_postgres();
    console.log(data);
    return processListPostgres(data);
  } else {
    throw Error("Invalid Source");
  }
}

/**
 * Retrieves array of eCR IDs from PostgreSQL database.
 * @returns A promise resolving to a NextResponse object.
 */
const list_postgres = async () => {
  const listFhir =
    "SELECT fhir.ecr_id, date_created, patient_name_last, patient_name_last, patient_birth_date, report_date, reportable_condition FROM fhir LEFT OUTER JOIN fhir_metadata on fhir.ecr_id = fhir_metadata.ecr_id order by date_created DESC";
  try {
    return await database.manyOrNone(listFhir);
  } catch (error: any) {
    console.error("Error fetching data:", error);
    if (error.message == "No data returned from the query.") {
      throw Error("No eCRs found");
    } else {
      throw Error(error.message);
    }
  }
};

/**
 * Retrieves array of eCRs and their metadata from S3.
 * @returns A promise resolving to a NextResponse object.
 */
const list_s3 = async () => {
  const bucketName = process.env.ECR_BUCKET_NAME;

  try {
    const command = new ListObjectsV2Command({
      Bucket: bucketName,
    });

    return await s3Client.send(command);
  } catch (error: any) {
    console.error("S3 GetObject error:", error);
    throw Error(error.message);
  }
};

/**
 * Processes a list of eCR data retrieved from Postgres.
 * @param responseBody - The response body containing eCR data from Postgres.
 * @returns - The processed list of eCR IDs and dates.
 */
export const processListPostgres = (responseBody: any[]): ListEcr => {
  return responseBody.map((object) => {
    return {
      ecrId: object.ecr_id || "",
      dateModified: object.date_created
        ? convertUTCToLocalString(
            formatDateTime(new Date(object.date_created!).toISOString()),
          )
        : "",
    };
  });
};

/**
 * Processes a list of eCR data retrieved from an S3 bucket.
 * @param responseBody - The response body containing eCR data from S3.
 * @returns - The processed list of eCR IDs and dates.
 */
export const processListS3 = (
  responseBody: ListObjectsV2CommandOutput,
): ListEcr => {
  responseBody.Contents?.sort(
    (a, b) => Number(b.LastModified) - Number(a.LastModified),
  );

  return (
    responseBody.Contents?.map((object) => {
      return {
        ecrId: object.Key?.replace(".json", "") || "",
        dateModified: object.LastModified
          ? convertUTCToLocalString(
              formatDateTime(new Date(object.LastModified!).toISOString()),
            )
          : "",
      };
    }) || []
  );
};
