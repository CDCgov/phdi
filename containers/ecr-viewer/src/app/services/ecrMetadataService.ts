import {
  formatAddress,
  formatContactPoint,
  formatDateTime,
} from "@/app/services/formatService";
import { PathMappings, evaluateData } from "@/app/view-data/utils/utils";
import { Bundle, Organization } from "fhir/r4";
import { evaluate } from "@/app/view-data/utils/evaluate";
import { DisplayDataProps } from "@/app/view-data/components/DataDisplay";
import { evaluateReference } from "@/app/services/evaluateFhirDataService";

export interface ReportableConditions {
  [condition: string]: {
    [trigger: string]: Set<string>;
  };
}

export interface ERSDWarning {
  warning: string;
  versionUsed: string;
  expectedVersion: string;
  suggestedSolution: string;
}

/**
 * Evaluates eCR metadata from the FHIR bundle and formats it into structured data for display.
 * @param fhirBundle - The FHIR bundle containing eCR metadata.
 * @param mappings - The object containing the fhir paths.
 * @returns An object containing evaluated and formatted eCR metadata.
 */
export const evaluateEcrMetadata = (
  fhirBundle: Bundle,
  mappings: PathMappings,
) => {
  const rrDetails = evaluate(fhirBundle, mappings.rrDetails);

  let reportableConditionsList: ReportableConditions = {};

  for (const condition of rrDetails) {
    let name = condition.valueCodeableConcept.coding[0].display;
    const triggers = condition.extension
      .filter(
        (x: { url: string; valueString: string }) =>
          x.url ===
          "http://hl7.org/fhir/us/ecr/StructureDefinition/us-ph-determination-of-reportability-rule-extension",
      )
      .map((x: { url: string; valueString: string }) => x.valueString);
    if (!reportableConditionsList[name]) {
      reportableConditionsList[name] = {};
    }

    for (let i in triggers) {
      if (!reportableConditionsList[name][triggers[i]]) {
        reportableConditionsList[name][triggers[i]] = new Set();
      }
      condition.performer
        .map((x: { display: string }) => x.display)
        .forEach((x: string) =>
          reportableConditionsList[name][triggers[i]].add(x),
        );
    }
  }

  const custodianRef = evaluate(fhirBundle, mappings.eicrCustodianRef)[0] ?? "";
  const custodian = evaluateReference(
    fhirBundle,
    mappings,
    custodianRef,
  ) as Organization;

  const eicrReleaseVersion = (fhirBundle: any, mappings: any) => {
    const releaseVersion = evaluate(fhirBundle, mappings.eicrReleaseVersion)[0];
    console.log(releaseVersion);
    if (releaseVersion === "2016-12-01") {
      return "R1.1 (2016-12-01)";
    } else if (releaseVersion === "2021-01-01") {
      return "R3.1 (2021-01-01)";
    } else {
      return releaseVersion;
    }
  };

  const fhirERSDWarnings = evaluate(fhirBundle, mappings.eRSDwarnings);
  let eRSDTextList: ERSDWarning[] = [];

  for (const warning of fhirERSDWarnings) {
    if (warning.code === "RRVS34") {
      eRSDTextList.push({
        warning:
          "Sending organization is using an malformed eRSD (RCTC) version",
        versionUsed: "2020-06-23",
        expectedVersion:
          "Sending organization should be using one of the following: 2023-10-06, 1.2.2.0, 3.x.x.x.",
        suggestedSolution:
          "The trigger code version your organization is using could not be determined. The trigger codes may be out date. Please have your EHR administrators update the version format for complete eCR functioning.",
      });
    } else if (warning.code === "RRVS29") {
      eRSDTextList.push({
        warning:
          "Sending organization is using an outdated eRSD (RCTC) version",
        versionUsed: "2020-06-23",
        expectedVersion:
          "Sending organization should be using one of the following: 2023-10-06, 1.2.2.0, 3.x.x.x.",
        suggestedSolution:
          "The trigger code version your organization is using is out-of-date. Please have your EHR administration install the current version for complete eCR functioning.",
      });
    }
  }

  const eicrDetails: DisplayDataProps[] = [
    {
      title: "eICR ID",
      toolTip:
        "Unique document ID for the eICR that originates from the medical record. Different from the Document ID that NBS creates for all incoming records.",
      value: evaluate(fhirBundle, mappings.eicrIdentifier)[0],
    },
    {
      title: "Date/Time eCR Created",
      value: formatDateTime(
        evaluate(fhirBundle, mappings.dateTimeEcrCreated)[0],
      ),
    },
    {
      title: "eICR Release Version",
      value: eicrReleaseVersion(fhirBundle, mappings),
    },
    {
      title: "EHR Manufacturer Model Name",
      value: evaluate(fhirBundle, mappings.ehrManufacturerModel)[0],
    },
    {
      title: "EHR Software Name",
      toolTip: "EHR system used by the sending provider.",
      value: evaluate(fhirBundle, mappings.ehrSoftware)[0],
    },
  ];

  const ecrCustodianDetails: DisplayDataProps[] = [
    {
      title: "Custodian ID",
      value: custodian?.identifier?.[0]?.value,
    },
    {
      title: "Custodian Name",
      value: custodian?.name,
    },
    {
      title: "Custodian Address",
      value: formatAddress(
        custodian?.address?.[0].line ?? [],
        custodian?.address?.[0].city ?? "",
        custodian?.address?.[0].state ?? "",
        custodian?.address?.[0].postalCode ?? "",
        custodian?.address?.[0].country ?? "",
      ),
    },
    {
      title: "Custodian Contact",
      value: formatContactPoint(custodian?.telecom).join("\n"),
    },
  ];

  return {
    eicrDetails: evaluateData(eicrDetails),
    ecrCustodianDetails: evaluateData(ecrCustodianDetails),
    rrDetails: reportableConditionsList,
    eRSDWarnings: eRSDTextList,
  };
};
