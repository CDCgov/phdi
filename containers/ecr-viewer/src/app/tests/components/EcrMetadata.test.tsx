import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import EcrMetadata from "../../view-data/components/EcrMetadata";
import React from "react";
import { DisplayDataProps } from "@/app/view-data/components/DataDisplay";
import { ReportableConditions } from "@/app/services/ecrMetadataService";

describe("ECR Metadata", () => {
  let container: HTMLElement;
  beforeAll(() => {
    const mockChildMethod = jest.fn();
    jest.spyOn(React, "useRef").mockReturnValue({
      current: {
        childMethod: mockChildMethod,
      },
    });
    const rrConditionsList: ReportableConditions = {
      "Disease caused by severe acute respiratory syndrome coronavirus 2(disorder)":
        {
          "Detection of SARS-CoV-2 nucleic acid in a clinical or post-mortem specimen by any method":
            new Set([
              "California Department of Public Health",
              "Los Angeles County Department of Public Health",
            ]),
          "Close contact in the 14 days prior to onset of symptoms with a confirmed or probable case of COVID-19 (Partially implemented as exposure with no timeframe parameters)":
            new Set(["Los Angeles County Department of Public Health"]),
          "COVID-19 (as a diagnosis or active problem)": new Set([
            "Los Angeles County Department of Public Health",
          ]),
        },
    };
    const eicrDetails: DisplayDataProps[] = [
      {
        title: "eICR Identifier",
        value: "1dd10047-2207-4eac-a993-0f706c88be5d",
      },
      {
        title: "Date/Time eCR Created",
        value: "2022-05-14T12:56:38Z",
      },
      { title: "eICR Release Version", value: "R1.1 (2016-12-01)" },
      {
        title: "EHR Software Name",
        value: "Epic - Version 10.1",
      },
      { title: "EHR Manufacturer Model Name", value: "Epic - Version 10.1" },
    ];
    const ecrCustodianDetails: DisplayDataProps[] = [
      {
        title: "Custodian ID",
        value: "1104202761",
      },
      {
        title: "Custodian Name",
        value: "Vanderbilt University Medical Center",
      },
      {
        title: "Custodian Address",
        value: "3401 West End Ave\nNashville, TN\n37203, USA",
      },
      {
        title: "Custodian Contact",
        value: "Work 1-615-322-5000",
      },
    ];

    container = render(
      <EcrMetadata
        eicrDetails={eicrDetails}
        eCRCustodianDetails={ecrCustodianDetails}
        rrDetails={rrConditionsList}
      />,
    ).container;
  });
  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });
  it("should pass accessibility test", async () => {
    expect(await axe(container)).toHaveNoViolations();
  });
});
