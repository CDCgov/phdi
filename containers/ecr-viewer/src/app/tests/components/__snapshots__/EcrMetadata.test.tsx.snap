// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ECR Metadata should match snapshot 1`] = `
<div>
  <div
    class="margin-top-0"
  >
    <div
      class="padding-bottom-3"
    >
      <div
        class="usa-summary-box__body"
      >
        <h4
          class="usa-summary-box__heading padding-y-105"
          id="rr-details"
        >
          RR Details
        </h4>
        <div
          class="usa-summary-box__text"
        >
          <table
            class="usa-table width-full src-components-Table-Table-module__fixed--Eq00W rrTable"
            data-testid="table"
          >
            <caption>
              Reportibility Summary
            </caption>
            <thead>
              <tr>
                <th
                  class="width-25p"
                >
                  <span
                    class="usa-tooltip"
                    data-testid="tooltipWrapper"
                  >
                    <div
                      aria-describedby="[object Object]"
                      class="usa-tooltip__trigger usa-tooltip"
                      data-testid="triggerElement"
                      position="top"
                      tabindex="0"
                      title=""
                    >
                      Reportable Condition
                    </div>
                    <span
                      aria-hidden="true"
                      class="usa-tooltip__body"
                      data-testid="tooltipBody"
                      id="[object Object]"
                      role="tooltip"
                      title="List of conditions that caused this eCR to be sent to your jurisdiction based on the rules set up for routing eCRs by your jurisdiction in RCKMS (Reportable Condition Knowledge Management System). Can include multiple Reportable Conditions for one eCR."
                    >
                      List of conditions that caused this eCR to be sent to your jurisdiction based on the rules set up for routing eCRs by your jurisdiction in RCKMS (Reportable Condition Knowledge Management System). Can include multiple Reportable Conditions for one eCR.
                    </span>
                  </span>
                </th>
                <th>
                  <span
                    class="usa-tooltip"
                    data-testid="tooltipWrapper"
                  >
                    <div
                      aria-describedby="[object Object]"
                      class="usa-tooltip__trigger usa-tooltip"
                      data-testid="triggerElement"
                      position="top"
                      tabindex="0"
                      title=""
                    >
                      RCKMS Rule Summary
                    </div>
                    <span
                      aria-hidden="true"
                      class="usa-tooltip__body"
                      data-testid="tooltipBody"
                      id="[object Object]"
                      role="tooltip"
                      title="Reason(s) that this eCR was sent for this condition. Corresponds to your jurisdiction's rules for routing eCRs in RCKMS (Reportable Condition Knowledge Management System)."
                    >
                      Reason(s) that this eCR was sent for this condition. Corresponds to your jurisdiction's rules for routing eCRs in RCKMS (Reportable Condition Knowledge Management System).
                    </span>
                  </span>
                </th>
                <th
                  class="width-25p"
                >
                  <span
                    class="usa-tooltip"
                    data-testid="tooltipWrapper"
                  >
                    <div
                      aria-describedby="[object Object]"
                      class="usa-tooltip__trigger usa-tooltip"
                      data-testid="triggerElement"
                      position="top"
                      tabindex="0"
                      title=""
                    >
                      Jurisdiction Sent eCR
                    </div>
                    <span
                      aria-hidden="true"
                      class="usa-tooltip__body"
                      data-testid="tooltipBody"
                      id="[object Object]"
                      role="tooltip"
                      title="List of jurisdictions this eCR was sent to. Can include multiple jurisdictions depending on provider location, patient address, and jurisdictions onboarded to eCR."
                    >
                      List of jurisdictions this eCR was sent to. Can include multiple jurisdictions depending on provider location, patient address, and jurisdictions onboarded to eCR.
                    </span>
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  rowspan="4"
                >
                  Disease caused by severe acute respiratory syndrome coronavirus 2(disorder)
                </td>
                <td
                  rowspan="2"
                >
                  Detection of SARS-CoV-2 nucleic acid in a clinical or post-mortem specimen by any method
                </td>
                <td>
                  California Department of Public Health
                </td>
              </tr>
              <tr>
                <td>
                  Los Angeles County Department of Public Health
                </td>
              </tr>
              <tr>
                <td
                  rowspan="1"
                >
                  Close contact in the 14 days prior to onset of symptoms with a confirmed or probable case of COVID-19 (Partially implemented as exposure with no timeframe parameters)
                </td>
                <td>
                  Los Angeles County Department of Public Health
                </td>
              </tr>
              <tr>
                <td
                  rowspan="1"
                >
                  COVID-19 (as a diagnosis or active problem)
                </td>
                <td>
                  Los Angeles County Department of Public Health
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="padding-bottom-1"
          />
          <h4
            class="usa-summary-box__heading padding-y-105"
            id="eicr-details"
          >
            eICR Details
          </h4>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                eICR Identifier
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                1dd10047-2207-4eac-a993-0f706c88be5d
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                Date/Time eCR Created
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                2022-05-14T12:56:38Z
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                eICR Release Version
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                R1.1 (2016-12-01)
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                EHR Software Name
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Epic - Version 10.1
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                EHR Manufacturer Model Name
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Epic - Version 10.1
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div
            class="padding-bottom-1"
          />
          <h4
            class="usa-summary-box__heading padding-y-105"
            id="eicr-custodian-details"
          >
            eICR Custodian Details
          </h4>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                Custodian ID
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                1104202761
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                Custodian Name
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Vanderbilt University Medical Center
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                Custodian Address
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                3401 West End Ave
Nashville, TN
37203, USA
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
          <div>
            <div
              class="grid-row"
            >
              <div
                class="data-title padding-right-1"
              >
                Custodian Contact
              </div>
              <div
                class="grid-col maxw7 text-pre-line"
              >
                Work 1-615-322-5000
              </div>
            </div>
            <div
              class="section__line_gray"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
